### ISSUE TITLE
Appservice IRC failed to deploy, please help!

### ISSUE THREAD

**Issue by Shillos** [spantaleev/matrix-docker-ansible-deploy#2108]
Hi! I don't know if this is a bug or something else, but when I run for the update after my Ansible deployment, I get this issue:

TASK [matrix-bridge-appservice-irc : Generate Appservice IRC registration-template.yaml] ***
fatal: [matrix.cyhub.org]: FAILED! => changed=false
  cmd: /usr/bin/env docker run --rm --name matrix-appservice-irc-gen --user=997:1002 --cap-docker.io/matrixdotorg/matrix-appservice-irc:release-0.35.0 -c 'node app.js -r -f /config/reg
  delta: '0:00:00.651527'
  end: '2022-09-16 15:46:52.344646'
  msg: non-zero return code
  rc: 1
  start: '2022-09-16 15:46:51.693119'
  stderr: |-
    WARNING: The requested image's platform (linux/amd64) does not match the detected host p
    exec /bin/bash: exec format error
  stderr_lines: <omitted>
  stdout: ''
  stdout_lines: <omitted>

I'm using Oracle Cloud Server with Ubuntu 20.04 and Ampere 1 4 core arch64 24gb ram 200gb space. It worked ok since the older version, and now it just doesn't work for no reason. Can anyone please help me out? I even asked in the matrix Ansible channel and no one gave me anything. Please help! Thanks!

**Comment by spantaleev** [comment]
[matrix-appservice-irc](https://github.com/matrix-org/matrix-appservice-irc) broke self-building with the `0.35.0` release, it seems.

I'm seeing some errors like this when self-building it:

```
Step 9/24 : RUN npm ci
 ---> Running in ab2a9fe52c80

> utf-8-validate@5.0.9 install /build/node_modules/utf-8-validate
> node-gyp-build

gyp ERR! find Python 
gyp ERR! find Python Python is not set from command line or npm configuration
gyp ERR! find Python Python is not set from environment variable PYTHON
gyp ERR! find Python checking if "python" can be used
gyp ERR! find Python - "python" is not in PATH or produced an error
gyp ERR! find Python checking if "python2" can be used
gyp ERR! find Python - "python2" is not in PATH or produced an error
gyp ERR! find Python checking if "python3" can be used
gyp ERR! find Python - "python3" is not in PATH or produced an error
gyp ERR! find Python 
gyp ERR! find Python **********************************************************
gyp ERR! find Python You need to install the latest version of Python.
gyp ERR! find Python Node-gyp should be able to find and use Python. If not,
gyp ERR! find Python you can try one of the following options:
gyp ERR! find Python - Use the switch --python="/path/to/pythonexecutable"
gyp ERR! find Python   (accepted by both node-gyp and npm)
gyp ERR! find Python - Set the environment variable PYTHON
gyp ERR! find Python - Set the npm configuration variable python:
gyp ERR! find Python   npm config set python "/path/to/pythonexecutable"
gyp ERR! find Python For more information consult the documentation at:
gyp ERR! find Python https://github.com/nodejs/node-gyp#installation
gyp ERR! find Python **********************************************************
gyp ERR! find Python 
gyp ERR! configure error 
gyp ERR! stack Error: Could not find any Python installation to use
gyp ERR! stack     at PythonFinder.fail (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/find-python.js:307:47)
gyp ERR! stack     at PythonFinder.runChecks (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/find-python.js:136:21)
gyp ERR! stack     at PythonFinder.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/find-python.js:179:16)
gyp ERR! stack     at PythonFinder.execFileCallback (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/find-python.js:271:16)
gyp ERR! stack     at exithandler (child_process.js:390:5)
gyp ERR! stack     at ChildProcess.errorhandler (child_process.js:402:5)
gyp ERR! stack     at ChildProcess.emit (events.js:400:28)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:283:12)
gyp ERR! stack     at onErrorNT (internal/child_process.js:472:16)
gyp ERR! stack     at processTicksAndRejections (internal/process/task_queues.js:82:21)
gyp ERR! System Linux 5.15.0-1017-oracle
gyp ERR! command "/usr/local/bin/node" "/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /build/node_modules/utf-8-validate
gyp ERR! node -v v14.20.0
gyp ERR! node-gyp -v v5.1.0
gyp ERR! not ok 
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! utf-8-validate@5.0.9 install: `node-gyp-build`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the utf-8-validate@5.0.9 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2022-09-17T06_59_43_169Z-debug.log
The command '/bin/sh -c npm ci' returned a non-zero code: 1
```

As a workaround, you can downgrade by putting `matrix_appservice_irc_version: 0.34.0` in your `vars.yml` file.

Opening an issue at the upstream repository would also be helpful to get it resolved for future versions.

**Comment by Shillos** [comment]
Thanks for your reply. I already did that, keeping the old version in my vars.yml when I noticed the odds. In my server, I already had python 3.8.10, but no node or NPM, so I installed NVM (Node Versioning Manager) just so it's easier to install future versions and always the LTS editions. OK, this is what I did. I installed the node-gyp, hoped that it worked I included the new commit line you added `matrix_appservice_irc_docker_image_name_prefix: "{{ 'localhost/' if matrix_appservice_irc_container_image_self_build else matrix_container_global_registry_prefix }}"` then I tried putting `matrix_appservice_irc_version: 0.35.0` run it again and had some odd issues, but when I changed the line from `matrix_appservice_irc_version: 0.35.0` to `matrix_appservice_irc_version: 'latest'` then it worked like a charm no errors found. Thanks!

**Comment by spantaleev** [comment]
All of that `node-gyp` and `python` stuff you see is stuff that happens during container image building. It happens in a container. Whatever `node-gyp` or `python` versions you install on the host will not have any effect.

**Comment by Shillos** [comment]
Having the same problems again after 0.35.1 release. https://pastebin.com/sggRxapC it's not building and I already have python and node-gyp and all installed, still not able to build with the latest version. This is getting too much, I bet others have the same problem.

**Comment by spantaleev** [comment]
Consider complaining about this upstream: https://github.com/matrix-org/matrix-appservice-irc

If Docker image building is broken there (for arm64, etc.), there's nothing we can do.

**Comment by Shillos** [comment]
Hi! Yes, I should actually send this to the creators on their upstream and I might do so for future problems on this issue, but I played around and found that your `matrix_appservice_irc_docker_image_name_prefix: "{{ 'localhost/' if matrix_appservice_irc_container_image_self_build else matrix_container_global_registry_prefix }}"` when building in the docker this had to be disabled in version 0.35.1 as I noticed something didn't add well, so it worked fine again. It might be they changed something and I do not know what. But again, thanks. I think this PR is done. :-)
